### PR TITLE
Incorporate the cipher suite in key derivation

### LINKED ISSUE #128: Incorporate cipher suite in key generation
```diff
-sframe_secret = HKDF-Extract("SFrame 1.0 Secret " + KID, base_key)
+sframe_secret = HKDF-Extract("SFrame 1.0 Secret " + KID + cipher_suite, base_key)
```

This would solve a couple of problems:

* We could remove the extra HKDF steps in generating encryption/authentication subkeys for the CTR+HMAC cipher suites.  Instead, we would just generate a longer key and split it at the appropriate point.  (We need the cipher suite baked into the KDF so that we get different keys for different tag lengths.)

* We would be slightly more resilient to `base_key` reuse, since you would have to reuse the same `base_key`, `KID`, **and** cipher suite.

The major drawback is that you need a cipher suite code point to use SFrame.  Obviously not a problem for the standard suites, but maybe people want to use unofficial suites.  Probably solvable with some vendor space.

### PR DESCRIPTION
**Pull request by bifurcation**
Fixes #128

### CODE REVIEW

File: draft-ietf-sframe-enc.md [hunk 1]
```diff
@@ -396,14 +396,21 @@ using HKDF {{!RFC5869}} as follows:
...
   return sframe_key, sframe_salt
 ~~~~~
 
-In the derivation of `sframe_secret`, the `+` operator represents concatenation
-of byte strings and the KID value is encoded as an 8-byte big-endian integer
-(not the compressed form used in the SFrame header).
+In the derivation of `sframe_secret`:
+* The `+` operator represents concatenation of byte strings.
+* The KID value is encoded as an 8-byte big-endian integer, not the compressed
+  form used in the SFrame header).
```
Comment (bifurcation) [inline draft-ietf-sframe-enc.md:411]:
```suggestion
  form used in the SFrame header.
```

File: draft-ietf-sframe-enc.md [hunk 2]
```diff
@@ -396,14 +396,21 @@ using HKDF {{!RFC5869}} as follows:
...
 def derive_key_salt(KID, base_key):
   sframe_secret = HKDF-Extract("", base_key)
-  sframe_key = HKDF-Expand(sframe_secret, "SFrame 1.0 Secret key " + KID, AEAD.Nk)
-  sframe_salt = HKDF-Expand(sframe_secret, "SFrame 1.0 Secret salt " + KID, AEAD.Nn)
+  sframe_key = HKDF-Expand(sframe_secret,
+                           "SFrame 1.0 Secret key " + KID + cipher_suite,
+                           AEAD.Nk)
+  sframe_salt = HKDF-Expand(sframe_secret,
+                            "SFrame 1.0 Secret salt " + KID + cipher_suite,
+                            AEAD.Nn)
```
Comment (martinthomson) [inline draft-ietf-sframe-enc.md:404]:
```suggestion
  info = "SFrame 1.0 Secret key " + KID + cipher_suite
  sframe_key = HKDF-Expand(sframe_secret, info, AEAD.Nk)
  sframe_salt = HKDF-Expand(sframe_secret, info, AEAD.Nn)
```

File: draft-ietf-sframe-enc.md [hunk 3]
```diff
@@ -546,18 +553,16 @@ using the authenticated counter mode of AES together with HMAC for
 authentication.  We use an encrypt-then-MAC approach, as in SRTP {{?RFC3711}}.
 
 Before encryption or decryption, encryption and authentication subkeys are
-derived from the single AEAD key using HKDF.  The subkeys are derived as
-follows, where `Nk` represents the key size for the AES block cipher in use,
-`Nh` represents the output size of the hash function, and `Nt` represents the
-size of a tag for the cipher in bytes (as in {{iana-cipher-suites}}):
+derived from the single AEAD key.  The overall length of the AEAD key is `Nk +
```
Comment (bifurcation) [inline draft-ietf-sframe-enc.md:556]: Give `Nk` a different name to avoid confusion with `Nk` for the overall AEAD.

**Comment by bifurcation** [comment]
Interim 2023-09-13:
* Lennox: Noted two editorial nits
* TODO(RLB): Fix nits and merge.

**Pull request merged**